torch/csrc/distributed/c10d/NCCLUtils.cpp

```diff
@@ -57,6 +57,57 @@ std::string ncclGetErrorWithVersion(ncclResult_t error) {
       getNcclVersion();
 }
 
+// Provides additional detail into NCCL error codes based on when these are
+// thrown in the NCCL codebase.
+std::string getNcclErrorDetailStr(
+    ncclResult_t error,
+    c10::optional<std::string> processGroupFailureReason /* = c10::nullopt */
+) {
+  // Prioritize failure reason provided by PG NCCL first, as it can abort
+  // communicators when it encounters collective timeouts, etc.
+  if (processGroupFailureReason != c10::nullopt) {
+    return *processGroupFailureReason;
+  }
+  std::string interpret;
+  std::string err;
+#ifdef ENABLE_NCCL_GET_LAST_ERROR
+  err = "\nLast error:\n" + std::string(ncclGetLastError(NULL));
+#endif
+  switch (error) {
+    case ncclUnhandledCudaError:
+      interpret = "ncclUnhandledCudaError: Call to CUDA function failed.";
+      break;
+    case ncclSystemError:
+      interpret =
+          "ncclSystemError: System call (e.g. socket, malloc) or external library call failed or device error. ";
+#ifndef NCCL_REMOTE_ERROR
+      // Before ncclRemoteError was created, unexpected remote disconnect was
+      // categorized as ncclSystemError
+      interpret += "It can be also caused by unexpected exit of a remote peer.";
+#endif
+      break;
+    case ncclInternalError:
+      interpret = "ncclInternalError: Internal check failed.";
+      break;
+    case ncclInvalidArgument:
+      interpret = "ncclInvalidArgument: Invalid value for an argument.";
+      break;
+    case ncclInvalidUsage:
+      interpret =
+          "ncclInvalidUsage: This usually reflects invalid usage of NCCL library.";
+      break;
+#ifdef NCCL_REMOTE_ERROR
```

Comment:
@crcrpar can you please retry https://github.com/pytorch/pytorch/pull/85782 with the above macro flag?
It does not apply to `ncclInProgress` though. 
Maybe we can land things one at a time? :)